### PR TITLE
Generate DOM for markdown links in help text

### PR DESCRIPTION
**Pull request by ebidel**
R: all

Just support in help text / descriptions.

screen shot 2017-04-07 at 1 57 59 pm

Will add a test once https://github.com/GoogleChrome/lighthouse/pull/1963 lands.

### CODE REVIEW

File: lighthouse-core/report/v2/report-renderer.js [hunk 1]
```diff
@@ -181,6 +182,36 @@ class ReportRenderer {
   }
 
   /**
+   * @param {string} text
+   * @return {!Element}
```
Comment (brendankenny) [inline lighthouse-core/report/v2/report-renderer.js]: `@return {!HTMLSpanElement}`. A function description here would be good too :)

File: lighthouse-core/report/v2/report-renderer.js [hunk 2]
```diff
@@ -181,6 +182,36 @@ class ReportRenderer {
...
+   */
+  _convertMarkdownLinksToElement(text) {
+    const element = this._createElement('span');
+
+    // Split on markdown links (e.g. [some link](https://...)).
+    const parts = text.split(/(\[(.*?)\]\((https?:\/\/.*?)\))/g);
```
Comment (brendankenny) [inline lighthouse-core/report/v2/report-renderer.js]: any reason to capture `fullMarkdownLink` if it's going to be unused?

File: lighthouse-core/report/v2/report-renderer.js [hunk 3]
```diff
@@ -181,6 +182,36 @@ class ReportRenderer {
...
+
+    while (parts.length) {
+      // Remove the same number of elements as there are capture groups.
+      // eslint-disable-next-line no-unused-vars
+      const [preambleText, fullMarkdownLink, linkText, linkHref] = parts.splice(0, 4);
```
Comment (brendankenny) [inline lighthouse-core/report/v2/report-renderer.js]: if you want to keep the larger capture group, you can also rename `fullMarkdownLink` to `_` to indicate that it's unused and drop the `// eslint-disable-next-line no-unused-vars`

File: lighthouse-core/test/report/v2/report-renderer-test.js [hunk 4]
```diff
@@ -26,10 +26,55 @@ const sampleResults = require('../../results/sample_v2.json');
...
 
 describe('ReportRenderer V2', () => {
-  describe('renderReport', () => {
-    const document = jsdom.jsdom(TEMPLATE_FILE);
-    const renderer = new ReportRenderer(document);
+  const document = jsdom.jsdom(TEMPLATE_FILE);
+  const renderer = new ReportRenderer(document);
+
+  describe('createElement', () => {
+    it('creates a simple element using default values', () => {
```
Comment (brendankenny) [inline lighthouse-core/test/report/v2/report-renderer-test.js]: what's new and what's not is very confusing when you move things around :P

File: lighthouse-core/test/report/v2/report-renderer-test.js [hunk 5]
```diff
@@ -26,10 +26,55 @@ const sampleResults = require('../../results/sample_v2.json');
...
+      assert.ok(clone.querySelector('.lighthouse-score'));
+    });
+
+    it('fails when template cannot be found', () => {
+      assert.throws(() => renderer._cloneTemplate('#unknown-selector'));
+    });
+  });
+
+  describe('_convertMarkdownLinksToElement', () => {
+    it('converts links', () => {
```
Comment (brendankenny) [inline lighthouse-core/test/report/v2/report-renderer-test.js]: A few more corner cases would be nice. With/without preamble text and text afterwards, etc

File: lighthouse-core/report/v2/report-renderer.js [hunk 6]
```diff
@@ -181,6 +182,36 @@ class ReportRenderer {
...
+      element.appendChild(this._document.createTextNode(preambleText));
+
+      // Append link if there are any.
+      if (linkText && linkHref) {
+        const a = this._createElement('a');
+        a.rel = 'noopener';
+        a.target = '_blank';
+        a.href = linkHref;
```
Comment (brendankenny) [inline lighthouse-core/report/v2/report-renderer.js]: do we want to verify the url? Maybe someday it would be good if we start using this in more places, but today it would start checking our help text URLs, which would be nice

File: lighthouse-core/report/v2/report-renderer.js [hunk 7]
```diff
@@ -181,6 +182,36 @@ class ReportRenderer {
...
+   * @return {!Element}
+   */
+  _convertMarkdownLinksToElement(text) {
+    const element = this._createElement('span');
+
+    // Split on markdown links (e.g. [some link](https://...)).
+    const parts = text.split(/(\[(.*?)\]\((https?:\/\/.*?)\))/g);
+
+    while (parts.length) {
+      // Remove the same number of elements as there are capture groups.
```
Comment (brendankenny) [inline lighthouse-core/report/v2/report-renderer.js]: I always forget that `splice` affects both creates a new array *and* modifies the original, so maybe change the comment to something like `// Pop off the same number of elements as there are capture groups.` to make it clear?

File: lighthouse-core/report/v2/report-renderer.js (hunk 3 above)
Comment (brendankenny) [inline lighthouse-core/report/v2/report-renderer.js]: ooh, can also do `const [preambleText, , linkText, linkHref]`, though don't know how our eslint would react

File: lighthouse-core/report/v2/report-renderer.js (hunk 7 above)
Comment (brendankenny) [inline lighthouse-core/report/v2/report-renderer.js]: you could also avoid the `splice` by doing something like
```js
let parts = text.split(/(\[(.*?)\]\((https?:\/\/.*?)\))/g);
while (parts.length) {
  const [preambleText, fullMarkdownLink, linkText, linkHref, ...rest] = parts;
  // ...
  parts = rest;
}
```
more explicit, though the extra assignment is kind of annoying

File: lighthouse-core/report/v2/report-renderer.js (hunk 1 above)
Comment (ebidel) [inline lighthouse-core/report/v2/report-renderer.js]: Is the type of element important We're not specifying subclass type for the other methods.

File: lighthouse-core/report/v2/report-renderer.js (hunk 6 above)
Comment (ebidel) [inline lighthouse-core/report/v2/report-renderer.js]: what do you mean by verify?

File: lighthouse-core/test/report/v2/report-renderer-test.js (hunk 4 above)
Comment (ebidel) [inline lighthouse-core/test/report/v2/report-renderer-test.js]: i like to keep you on your toes.

File: lighthouse-core/report/v2/report-renderer.js (hunk 1 above)
Comment (brendankenny) [inline lighthouse-core/report/v2/report-renderer.js]: > Is the type of element important We're not specifying subclass type for the other methods.

not really, though it doesn't hurt. I'm more thinking for documentation purposes it's nice to have the information that your linkified text is coming back in a `<span>`.

File: lighthouse-core/test/report/v2/report-renderer-test.js (hunk 5 above)
Comment (ebidel) [inline lighthouse-core/test/report/v2/report-renderer-test.js]: done

File: lighthouse-core/report/v2/report-renderer.js (hunk 3 above)
Comment (ebidel) [inline lighthouse-core/report/v2/report-renderer.js]: i just removed the extra group. no longer needed.

File: lighthouse-core/report/v2/report-renderer.js (hunk 2 above)
Comment (ebidel) [inline lighthouse-core/report/v2/report-renderer.js]: nope done.

File: lighthouse-core/report/v2/report-renderer.js (hunk 6 above)
Comment (brendankenny) [inline lighthouse-core/report/v2/report-renderer.js]: > what do you mean by verify?

sorry, I meant like check that it's a valid url. `a.href = new URL(linkHref).href` or whatever

File: lighthouse-core/report/v2/report-renderer.js (hunk 7 above)
Comment (ebidel) [inline lighthouse-core/report/v2/report-renderer.js]: updated the comment.

File: lighthouse-core/report/v2/report-renderer.js (hunk 6 above)
Comment (ebidel) [inline lighthouse-core/report/v2/report-renderer.js]: done

**Comment by ebidel** [comment]
PTAL

File: lighthouse-core/report/v2/report-renderer.js [hunk 8]
```diff
@@ -181,6 +182,39 @@ class ReportRenderer {
...
+      if (linkText && linkHref) {
+        const a = this._createElement('a');
+        a.rel = 'noopener';
+        a.target = '_blank';
+        a.textContent = linkText;
+        try {
+          a.href = (new URL(linkHref)).href;
+          element.appendChild(a);
+        } catch (e) {
+          element.appendChild(this._document.createTextNode(linkText));
```
Comment (brendankenny) [inline lighthouse-core/report/v2/report-renderer.js]: should we fail silently here? Seems like a big enough mistake that it should complain loudly and end up in `_renderException`

Comment (ebidel) [inline lighthouse-core/report/v2/report-renderer.js]: I think it's better to fail silently here. If a link is busted for some reason, it shouldn't block the whole report from being generated.

Comment (brendankenny) [inline lighthouse-core/report/v2/report-renderer.js]: but why is the link busted? Right now, at least, they're all coming from helpText. An error in one of those should be fixed immediately

Comment (ebidel) [inline lighthouse-core/report/v2/report-renderer.js]: done. 

FWIW, b/c of the regex, the only time we'd get an invalid URL is if someone wrote `https://`. or used something crazy in help text like `[boom](🍣🍺)`.

**Comment by ebidel** [comment]
This ready. Let's get it landed so other PRs can proceed.

**Comment by patrickhulce** [comment]
let's merge this so we can rebase #1987 and #1992 💥
